### PR TITLE
BBA/BuiltIn: Fix random network errors

### PR DESCRIPTION
**Pull request by sepalani**
This PR fixes Kirby Air Ride LAN mode not working for me. I had several issues. The first one being, for some weird reasons Dolphin was switching back and forth between my WiFi and VM networks. The second one, there was a timing issue where an instance of Dolphin send an UDP packet to the other one but it didn't have enough time to open the UDP port.

I addressed the first issue by binding the socket to the right network interface. The second issue is addressed by reserving the UDP ports during the initialisation rather than waiting to receive the right SSDP packet.

@JMC47 
Does it solve your connection issue?

@schthack @nolrinale
You might want to check that this PR doesn't introduce regression regarding the PSO games.

_One last thing regarding Windows 10. Testing this was a **nightmare**, I think the issue is Windows not my router but I had to do more thorough testing to confirm that. Often, my Windows 10 laptop blacklists (temporarily) my other devices (broadcasting UDP packets via Dolphin, namely Android and Linux) even with most protections (AV, Firewall, ...) turned off and then drop all packets from them. I had to manually send packets from my Windows 10 to the other devices to removed them (temporarily... it seems ><') from this blacklist. This is a standard Windows 10 Home edition install, I'm unsure what is causing this network issue but when sorted and Dolphin issues addressed, the BuiltIn adapter worked as expected._

Ready to be reviewed & tested.

### CODE REVIEW

**Comment by shuffle2** [comment]
>  Often, my Windows 10 laptop blacklists (temporarily) my other devices (broadcasting UDP packets via Dolphin, namely Android and Linux) even with most protections (AV, Firewall, ...) turned off and then drop all packets from them

I have to say, it seems pretty unlikely. Maybe dolphin is sending somehow incorrect packets, or windows chooses to route them to another interface (a VM adapter for instance). Should debug with wireshark or check routing tables, etc :)

**Comment by JMC47** [comment]
@shuffle2 it happened to me as well, where my laptop and my desktop would not connect, but my android phone could see both.

**Comment by sepalani** [comment]
> > Often, my Windows 10 laptop blacklists (temporarily) my other devices (broadcasting UDP packets via Dolphin, namely Android and Linux) even with most protections (AV, Firewall, ...) turned off and then drop all packets from them
> 
> I have to say, it seems pretty unlikely. Maybe dolphin is sending somehow incorrect packets, or windows chooses to route them to another interface (a VM adapter for instance). Should debug with wireshark or check routing tables, etc :)

I had Wireshark running on most devices and could see packets being broadcasted but not being received by the Windows 10 host. I also had a terminal on both Linux and Android to send ICMP pings (which was enabled on Windows) and I got back a "Destination host unreachable" error when the packets weren't received (no trace of the ICMP packets on Windows' Wireshark, I listened on all my network interfaces, just in case). As soon as my Windows contacted the other devices by itself (i.e. me sending a packet to the other devices using ping or netcat), the packets from the other devices got through again.

When I have more time, I'll enable more log and check if the error comes from Windows, Dolphin's forged packets. Might be a routing issues as you suggested.

**Comment by JMC47** [comment]
So this fixes the main issue of not being able to connect to real GameCube.  However, it results in the "port 76c is already in use" bug every time I connect to LAN.  It seems to connect anyway, just thought I'd note it.

**Comment by sepalani** [comment]
@JMC47 
Are you getting this error on Windows? If so, did you disable Windows' SSDP Discovery service? Is there another instance of Dolphin running and holding this port, otherwise?

**Comment by JMC47** [comment]
There's no other Dolphin's running.  I don't know what the Windows SSDP Discovery service is.  All of the games are working now, but that error does pop up every single time.

image

**Comment by JMC47** [comment]
Android crashes on this Pull Request *when booting* BBA titles.  Other titles boot fine.  It's weird because it's not even getting to the point where it tries to connect to LAN.

**Comment by JMC47** [comment]
I tried to do more testing, but founds things were much more touchy when trying to run 3/4 player runs, and then Windows did that shit where it will no longer see other devices, so I need to stop for a while.

**Comment by JMC47** [comment]
So I've done an experiment after waiting ~30 mins, where I try to connect two Windows PCs, Android, and a real GameCube.

Once the two Windows PCs are connected, the real GameCube seems to not be able to connect on Mario Kart: Double Dash!!  If I throw my Android device into the mix, the Real GameCube does see it (using master, as this PR crashes it. ) but Android also sees my real GameCubes and everything locks up.

**Comment by JMC47** [comment]
I've mostly been testing Mario Kart Double Dash a bit more as it seems to be the easiest, will move onto the other games to see if there is any difference in behavior.

Mario Kart Double Dash:  Windows cannot see the GameCube at all now.  Android is still crashing when using this PR, so I had to use Master on Android.  Master on Android can connect to my Windows machines *as long as I don't have a real GameCube searching*.  If I have a real GameCube searching, Android tries to connect to all 3 and we end up with a lockup since the Windows machines fail to connect to the GameCube.

Kirby Air Ride:  The GameCube IMMEDIATELY sees my Windows machines, but my Windows Desktop cannot see the GameCube and the connection fails.

1080 Snowboarding:  Connection between emulators is fine.  I need to find my disc to test on console and I honestly have no idea which container it is in.

**Comment by JMC47** [comment]
Sorry if this testing was a bit overload.  I didn't actually have my GameCube + BBA setup for the initial Pull Request (and I thought it was just for PSO - I didn't know it'd work for LAN games at a glance.)

**Comment by sepalani** [comment]
@JMC47
Do you have a stacktrace or anything that can help regarding the Android crash? I'm testing this PR with Android Studio and I haven't faced any, atm. I suspected a race condition so I added mutexes but it doesn't seem that's your issue. Don't worry about the testing, feel free to provide as much feedback as you want, it's very helpful.

I found other issues so I'm switching this PR to a draft due to its unstable nature. I've mainly tested Kirby Air Ride and I know what triggers a peer to be found. Since it's using UPnP, it expects a SSDP NOTIFY packet when a console joins (`NTS: ssdp:alive`) or leaves (`NTS: ssdp:byebye`) the LAN. If this packet is lost in-transit or by Dolphin, the game won't notice a peer has join/left the LAN. 

Regarding my testing, my results differ and seem to be performance-based. A working build and scenario might not work again when:
 * I launch Dolphin with a debugger (Visual Studio / Android Studio)
 * I capture the network using Wireshark
 * I watch a long video on Youtube

Now, the issues I faced so far:
 * The wrong network interface is used without this PR.
 * It seems UDP sockets are blocking despite the non-blocking call before bind.
 * The BuiltIn adapter doesn't seem to keep up when there are too many packets to handle.
   * Some packets are lost by Dolphin I think (according to the Network view's PCAP dump of the BBA traffic)
     * The PCAP dump from Wireshark shows the missing packets
     * Further testing with netcat listening shows that's not a binding address issue 
   * The way sockets are read seems unfair and tends to prefer the oldest sockets rather than the new ones

I also think they are also unnecessary hack that we can get rid of with a bit of work. This might please Windows not to drop some packets on top of that. I will split this PR into smaller ones to address these issues in order not to introduce regressions.

**Comment by JMC47** [comment]
I can get a stacktrace on Android when I reinstall Android Studio; I ran out of space on the harddrive it was on and never bothered to reinstall it on the other one.

**Comment by ANR2ME** [comment]
On some Win10 version there seems to be an issue of corruption https://social.technet.microsoft.com/Forums/en-US/965e107e-d9b0-4240-ac3f-74797c91b476/unable-to-send-udp-packets-larger-than-the-mtu-with-windows-build-1809-using-c-udpclient?forum=win10itpronetworking
> I encountered this problem recently at my company when we updated a number of our systems to 1809 and 1903. I did a bunch of debugging and analysis and I think I know what the problem is and how to work around it.
> 
> Problem is that the UDP packets larger than MTU in size (larger than 1472 bytes of max payload), will be corrupted and dropped by the network stack. 

> My conclusion is that this optimization (UDP Checksum Offload) was broken in the 1809 release and continues to remain broken in 1903. To work around it for now, the optimization needs to be disabled (seems to be enabled by default). This can be done in the Device Manager. Find your network adapter card in the list and go into Advanced tab. We only deal with IPv4, but IPv6 might also be affected, so you might need to disable both.

Not sure whether this issue have been fixed on newer Windows or not tho.

**Comment by shuffle2** [comment]
I have to wonder what people expect sending packets larger than MTU would do. It is not well defined and essentially up to the end-user to ensure their network and hardware can handle the traffic.

**Comment by nolrinale** [comment]
Testing with PSO on Xcode I just keep getting `Warning: Partial sends might not be handled properly.` every single time an action or movement happens on screen, the game works but its very concerning it keeps throwing this error in every single frame. 

You might want to check the `Socket::Status TcpSocket::send(const void* data, std::size_t size)` function to find the problem

**Comment by nolrinale** [comment]
Upon further inspection with the latest build from master to compare, I keep getting the warning as well, since its related to BBA maybe you can take a look at it too?

**Comment by JMC47** [comment]
I don't really have a better place to discuss this, but I was thinking about opening up a pull request to clear up some naming issues and wondered some opinions.  BBA Built In really comes off as weird when *technically* all the BBA options are built in.  I was thinking that maybe we should do it more like DSP.

BBA HLE
BBA LLE (TAP)
BBA LLE (XLink Kai)

I think that would give a clearer picture of the situation.  But, the reason why I haven't gone ahead with that is that I wanted to make sure that it makes sense to call this BBA HLE.  I *think* it is, as I thought the whole reason the TAP adapter was required was because it was handled through LLE.

**Comment by sepalani** [comment]
@nolrinale 
Are you able to bissect which version of Dolphin introduced this regression? Feel free to open an issue on the issue tracker too: https://dolp.in/i

@JMC47 
I think that's a good idea and make more sense that way. It's also clearly HLE'ing a network stack so I'm fine with these names.

**Comment by nolrinale** [comment]
I wasn't unable to pinpoint from where it began to appear but I figured since the name of this PR is `Fixing random network errors` related to BBA and this error _is_ from the BBA it should be included over here as well rather than creating a brand new issue for it on the issue tracker.

**Comment by nolrinale** [comment]
The issue has been opened
https://bugs.dolphin-emu.org/issues/12996

**Comment by schthack** [comment]
sorry for the late reply but for the Lan games on windows the SSDPSRV service must be disabled for the game to have an easy connection else the discovery wont work as windows use the port 1900

For pso as long as you dont toutch UDP and break DNS it will be fine, pso is tcp only

**Comment by nolrinale** [comment]
> 

In PSO's game settings you can force UDP to be always On, Off or Auto, isn't there a risk it could crash if you enable this option in particular for example by setting it to "Always On"?

**Comment by schthack** [comment]
@nolrinale Depends of the server, but for stability and prevent harmfull hack the UDP is disabled on most server and the game has to use TCP only

@sepalani i will look at the code, mabe we lost something in the refactoring for those game, but i remember Kirby to be one of the most anoying game to get working

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 1]
```diff
@@ -454,33 +458,28 @@ void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(const Common::UDPPacket&
...
       PanicAlertFmt(
           "Port {:x} is already in use, this game might not work as intented in LAN Mode.",
           htons(udp_header.source_port));
-      if (ref->udp_socket.bind(sf::Socket::AnyPort) != sf::Socket::Done)
+      if (ref->udp_socket.Bind(sf::Socket::AnyPort) != sf::Socket::Done)
       {
         ERROR_LOG_FMT(SP1, "Couldn't open UDP socket");
         return;
       }
-      if (ntohs(udp_header.destination_port) == 1900)
```
Comment (schthack) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:467]: this whole section has to go after the IF part at line 461, should be after 484 basicly... this is a bug from my code, duno why it ended there

File: Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp [hunk 2]
```diff
@@ -454,33 +458,28 @@ void CEXIETHERNET::BuiltInBBAInterface::HandleUDPFrame(const Common::UDPPacket&
...
+      if (ntohs(udp_header.destination_port) == 1900 && udp_header.length > 150)
```
Comment (schthack) [inline Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp:471]: if (ntohs(udp_header.destination_port) == 1900 && ntohs(udp_header.length) > 150)

**Comment by sepalani** [comment]
> sorry for the late reply, but for the Lan games on windows the SSDPSRV service must be disabled to allow a painless connection. else the discovery wont work as windows use the port 1900 and some game like Kirby need both end to receive a discovery notification...

JMC47 got it working without disabling the Windows SSDP discovery service. In fact, Windows listens on the SSDP port with all interfaces except for the "any" (i.e. 0.0.0.0) interface. On top of that, if we use the socket options `SO_REUSEADDR`/`SO_REUSEPORT` we can bind on the same address/port as the Windows service and still receive the multicast messages we subscribed to. By HLE'ing multicast properly we might be able to get rid of that issue.

@JMC47 
Do you still have the Android crash with these PRs?
 - https://github.com/dolphin-emu/dolphin/pull/10918
 - https://github.com/dolphin-emu/dolphin/pull/10917

**Comment by ANR2ME** [comment]
As i remembered when binding to a port already bound by other program with SO_REUSEPORT, both programs will share the data, so if one of them already received the packet, the other one won't be able to receive that same packet.

You may want to test for this behavior, hopefully i remembered it wrong (was having an issue with SO_REUSEPORT while testing AdhocServer on multiple instance of PPSSPP long ago)

**Comment by sepalani** [comment]
Closed in favour of https://github.com/dolphin-emu/dolphin/pull/10921.

**Pull request closed without merging**